net/wgpeerselector/files/usr/bin/wgpeerselector

```diff
@@ -0,0 +1,501 @@
+
+local function log(level, message)
+	if _G.is_verbose then
+		local prefix = {
+			[syslog.LOG_ERR]  = "Error: ",
+			[syslog.LOG_CRIT] = "Critical: ",
+			[syslog.LOG_INFO] = "Info: ",
+			[syslog.LOG_WARNING] = "Warning: ",
+		}
+		io.stderr:write(prefix[level]..message .. '\n')
+	end
+
+	syslog.syslog(level, message)
+end
+
+local function execute(command)
+	local f = io.popen(command)
+	if not f then
+		return nil
+	end
+
+	local data = f:read('*a')
+	f:close()
+	return data
+end
+
+local function sleep(secs)
+	-- This function calls uloop.run() for 'secs' seconds, so we can handle i/o
+	-- during that time. But be aware, that an fd should be registered to uloop,
+	-- because otherwise this function uses 100% cpu. This might be a bug in
+	-- uloop.
+	local timer = uloop.timer(function() uloop.cancel() end)
+	timer:set(secs*1000)
+	uloop.run()
+end
+
+local WGPeer = {}
+
+function WGPeer:new(o)
+	o = o or {}   -- create object if user does not provide one
+	setmetatable(o, self)
+	self.__index = self
+	-- some defaults
+	o.rx_bytes = 0
+	o.tx_bytes = 0
+	o.latest_handshake = 0
+	o.established_at = 0
+	-- terminology:
+	---> endpoint (maybe) needs dns resolution
+	---> remote is already resolved
+	o.next_remotes = {}
+	return o
+end
+
+function WGPeer:get_next_remote()
+	local ipv4_endpoint = string.match(self.endpoint, "^(%d+%.%d+%.%d+%.%d+:%d+)$")
+	if ipv4_endpoint then
+		return ipv4_endpoint
+	end
+
+	local ipv6_endpoint = string.match(self.endpoint, "^(%[[%x:]+%]:%d+)$")
+	if ipv6_endpoint then
+		return ipv6_endpoint
+	end
+
+	-- We cycle round robin through the dns responses. Therefore we keep the
+	-- results of the dns responses in self.next_remotes and pop one randomly
+	-- each time. If self.next_remotes is empty, do dns resolution again.
+	if #self.next_remotes < 1 then
+		-- do dns resolution
+		local host, port = string.match(self.endpoint, "^([%w-.]+):(%d+)$")
+		if not host or not port then
+			log(syslog.LOG_ERR, "Endpoint '"..self.endpoint.."' is not a valid endpoint.")
+			return nil
+		end
+
+		local res, _, errcode = sys_sock.getaddrinfo(host, nil, { socktype = sys_sock.SOCK_DGRAM })
+
+		if errcode then
+			log(syslog.LOG_WARNING, "DNS resolution of " .. self.endpoint .. " failed: error code " .. tostring(errcode))
+			return nil
+		end
+
+		for i = 1, #res do
+			if string.match(res[i].addr, ":") then -- IPv6
+				table.insert(self.next_remotes, '['..res[i].addr..']:'..port)
+			else -- IPv4
+				table.insert(self.next_remotes, res[i].addr..':'..port)
+			end
+		end
+	end
+
+	return table.remove(self.next_remotes, math.random(#self.next_remotes))
+end
+
+function WGPeer:install_to_kernel()
+	local remote = self:get_next_remote()
+	if not remote then
+		return false
+	end
+
+	local command = "wg set '" .. self.iface .. "' peer '" .. self.public_key .. "' " ..
+	                "allowed-ips '" .. table.concat(self.allowed_ips, ",") .. "' " ..
+	                "endpoint '" .. remote .. "' " ..
```

Comment:
`wg set ... endpoint ...` will resolve dns itself, so maybe you don't need get_next_remote? Unless you're really intent on doing some kind of round-robin multi homing, but usually that doesn't make sense for wireguard, since you rarely want to share private keys.

Reply:
This has mainly two reasons:
1. Better control of the error handling in case of dns failures.
2. To make connection attempts with IPv4 and IPv6 in alternation (IPv6, IPv4, IPv6, IPv4, IPv6, ...).